add docs on new testIsolation configuration option. (#4634)
cypress-io/cypress-documentation: content/guides/core-concepts/writing-and-organizing-tests.md

})
```

### Test Isolation

<Alert type="success">

<Icon name="check-circle" color="green"></Icon> **Best Practice:** Clean up
state **before** tests run.

</Alert>

> Test isolation is the practice of resetting the state _before_ each test.

Reviewer: ```suggestion
Test isolation is the practice of resetting application state _before_ each test.
```

> Cleaning up state ensures the operations of one test does not affect another
> test downstream. When writing tests, the goal for the test is reliably pass when
> ran standalone or in a randomized order. If you have set up your tests in a way
> that requires the state of a previous test, this could potentially break your
> tests.

Reviewer 2: ```suggestion
Cleaning up state ensures the operations of one test does not affect another
test downstream. The goal for each test should be to reliably pass when
run by itself or by having order independence when run with adjacent tests. 
Having tests that require the state of a previous test could potentially cause deterministic test failures.
```

Reviewer: My suggestion, incorporating the suggestion above:

```suggestion
Cleaning up state ensures that the operation of one test does not affect another
test later on. The goal for each test should be to reliably pass whether
run in isolation or consecutively with other tests. Having tests that depend on the state of an earlier test can potentially cause nondeterministic test failures.
```

Cypress supports two modes of test isolation, `default` and `strict`.

#### Default Mode

When in `default` mode, Cypress handles resetting the state for:

- [aliases](/api/commands/as)
- [cookies](/api/commands/clearcookies)
- [clock](/api/commands/clock)
- [intercepts](/api/commands/intercepts)
- [localStorage](/api/commands/clearlocalstorage)
- [routes](/api/commands/route)
- [sessions](/api/commands/session)
- [spies](/api/commands/spy)
- [stubs](/api/commands/stub)
- [viewport](/api/commands/viewport)

#### Strict Mode

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

> Strict mode is currently experimental and can be enabled by setting
> the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
> to `true` in the Cypress config. This is the default test isolation behavior
> when using the `experimentalSessionAndOrigin` experiment.
> 

Reviewer 2: Can `testIsolation` at the suite level be set to `default` with `experimentalSessionAndOrigin` enabled? I feel like that might be implied here when we might not want it to be.

Is there a way to turn on test isolation without enabling `experimentalSessionAndOrigin` globally? I see below that test isolation is configurable at the suite level, but what about globally?
```suggestion
Strict mode is currently experimental and can be enabled by setting
the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config. This is the default test isolation behavior
when using the `experimentalSessionAndOrigin` experiment.

```

```suggestion
Strict mode is currently experimental and can be enabled by setting
the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config. Test isolation behavior is set to `strict`
when using the `experimentalSessionAndOrigin` experiment.

```

Author: > Can testIsolation at the suite level be set to default with experimentalSessionAndOrigin enabled? I feel like that 
might be implied here when we might not want it to be.

Yes, that is the whole point of this ENV is to give users the ability to remove the clear page once Sessions/Origin goes GA and we do this by default which will cause some people to have to rewrite their tests.

> Is there a way to turn on test isolation without enabling experimentalSessionAndOrigin globally? I see below that test isolation is configurable at the suite level, but what about globally?

Current test isolation is always on in Cypress. The key different it test isolation today as-is vs the additional test isolation (clear page) forced when experimentalSessionAndOrigin=true.  Currently if users set `testIsolation=strict` without experimentalSessionAndOrigin=true, we do nothing. Might be worth exploring adding this in as a first-class opt-in though...

Because there is only one mode by default, I didn't make  it a global config in the linked pr, but specified to [e2e config object](https://github.com/cypress-io/cypress-documentation/pull/4634/files#diff-2ddd6ad3e6f0e14719d47ad1df70d747c592bd20195c7971b23f0669b2547a95R79). I pending feedback, I should update the config change to be on the e2e config, not the global config. good catch

</Alert>

When in `strict` mode, Cypress handles resetting the state for everything
outlined above for `default` mode, in addition to clearing the page by visiting
`about:blank` before each test. This clears the current application state by
forcing the tester to re-build the application's state for the test to ensure

> the test can reliably pass when ran standalone or in a randomized order.

Reviewer: > This clears the current application state by forcing the tester to re-build the application's state for the test

This is not quite true.

Setting the page to `about:blank` clears non-persistent browser state only. It does not affect localStorage or cookies (which is fine because Cypress already does that) but it also has no effect on any backend state. So if they are running their frontend against a stateful server of some kind, e.g. a REST API connected to a database, this still contains application state and still needs to be reset before every test if complete test isolation is to be achieved.

Not sure how much of this we need to include here, but we definitely shouldn't imply that blanking the page will clear 100% of application state.

Author: gosh you are right...this is how we've been describing it to try to clarify and categorize this why we need to visit about:blank but theres def more than just the rendered dom that matters to fully clear the application. thanks for calling this out!

### Test Isolation

<Alert type="success">

<Icon name="check-circle" color="green"></Icon> **Best Practice:** Clean up
state **before** tests run.

</Alert>

Test isolation is the practice of resetting application state _before_ each
test.

Cleaning up state ensures that the operation of one test does not affect another
test later on. The goal for each test should be to reliably pass whether run in
isolation or consecutively with other tests. Having tests that depend on the
state of an earlier test can potentially cause nondeterministic test failures.

Cypress supports two modes of test isolation, `lax` and `strict`.

#### Lax Mode

When in `lax` mode, Cypress handles resetting the state for:

- [aliases](/api/commands/as)
- [cookies](/api/commands/clearcookies)
- [clock](/api/commands/clock)
- [intercepts](/api/commands/intercepts)
- [localStorage](/api/commands/clearlocalstorage)
- [routes](/api/commands/route)
- [sessions](/api/commands/session)
- [spies](/api/commands/spy)
- [stubs](/api/commands/stub)
- [viewport](/api/commands/viewport)

#### Strict Mode

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

> Strict mode is currently experimental and can be enabled by setting

Reviewer 2: Not sure if this a good suggestion since it seems grammatically out of place or if it keeps the strict reference consistent through the markdown? 🤷🏻 

```suggestion
`strict` mode is currently experimental and can be enabled by setting
```

the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config. This is the default test isolation behavior
when using the `experimentalSessionAndOrigin` experiment.

</Alert>

When in `strict` mode, Cypress handles resetting the state for everything
outlined above for `lax` mode, in addition to clearing the page by visiting
`about:blank` before each test. This clears the dom's state and non-persistent
browser state. This forces you to re-visit your application and performing the

> serious of interactions needed to build the dom state and browser state for the

Reviewer 2: ```suggestion
series of interactions needed to build the dom state and browser state for the
```

> browser state. This forces you to re-visit your application and performing the
> series of interactions needed to build the dom state and browser state for the
> tests so it can reliably pass when ran standalone or in a randomized order.

Reviewer 2: ```suggestion
browser state. This forces you to re-visit your application and perform the
series of interactions needed to build the dom and browser state so the tests can reliably pass when run standalone or in a randomized order.
```

> The test isolation mode is a global configuration and can be override at the
> describe level with the
> [`testIsolation`](./guides/references/configuration#global).

Reviewer 2: ```suggestion
The test isolation mode is a global configuration and can be overridden at the
`describe` level with the
[`testIsolation`](./guides/references/configuration#global) option.
```
